contracts/MultiRewards.sol

```diff
@@ -376,64 +393,90 @@ library SafeMath {
     }
 }
 
-contract MultiRewards is ReentrancyGuard, Pausable {
-    using SafeMath for uint256;
+contract MultiFeeDistribution is
+    Initializable,
+    PausableUpgradeable,
+    OwnableUpgradeable
+{
     using SafeERC20 for IERC20;
 
     /* ========== STATE VARIABLES ========== */
+    struct RewardData {
+        uint256 amount;
+        uint256 lastTimeUpdated;
+        uint256 rewardPerToken;
+    }
 
-    struct Reward {
-        address rewardsDistributor;
-        uint256 rewardsDuration;
-        uint256 periodFinish;
-        uint256 rewardRate;
-        uint256 lastUpdateTime;
-        uint256 rewardPerTokenStored;
+    struct UserData {
+        uint256 tokenAmount;
+        uint256 lastTimeUpdated;
+        uint256 tokenClaimable;
+        mapping(address => uint256) rewardPerToken;
     }
-    IERC20 public stakingToken;
-    mapping(address => Reward) public rewardData;
+    address public stakingToken;
+    mapping(address => RewardData) public rewardData;
     address[] public rewardTokens;
 
-    // user -> reward token -> amount
-    mapping(address => mapping(address => uint256)) public userRewardPerTokenPaid;
-    mapping(address => mapping(address => uint256)) public rewards;
+    /// @notice address => RPT
+    mapping(address => UserData) public userData;
+    mapping(address => mapping(address => uint256)) public claimable;     /// @notice rewardToken => user => claimable amount
 
-    uint256 private _totalSupply;
-    mapping(address => uint256) private _balances;
+    uint256 private totalStakes;
+    // mapping(address => uint256) private _balances; // the stakingToken balances are encoded in userData mapping i.e. userData[account].tokenAmount
 
     /* ========== CONSTRUCTOR ========== */
 
-    constructor(
-        address _owner,
-        address _stakingToken
-    ) public Owned(_owner) {
-        stakingToken = IERC20(_stakingToken);
+    function initialize(
+        address[] memory _rewardTokens
+    ) public initializer {
+        for (uint i; i < _rewardTokens.length; i ++) {
+            if (_rewardTokens[i] == address(0)) revert InvalidBurn();
+            rewardTokens.push(_rewardTokens[i]);
+        }
+
+        __Pausable_init();
+        __Ownable_init();
     }
 
     function addReward(
-        address _rewardsToken,
-        address _rewardsDistributor,
```

Comment:
don't really understand the point of the `_rewardsDistributor` or `_rewardsDuration` tbh.

Reply:
I think Curve's contract allows rewardDistributors to control flow of rewards. They send rewards into the contract and they can choose when to stop distribution. In our case (and in Gamma case), vault is essentially a distributor, so all this functionality is not needed